File: c/cat.c
#include <stdio.h>
int main() {
    int exit_code = 0; // success
    char buf[1024];
    for(;;) {
        size_t count = fread(buf, sizeof(char), sizeof(buf), stdin);
        if(count > 0) {
            size_t written = fwrite(buf, sizeof(char), count, stdout);
            if(written != count) {
                fprintf(stderr, "Failed to write to stdout\n");
                exit_code = 1;
                break;
            }
        } else {
            if(ferror(stdin)) {
                fprintf(stderr, "Failed to read from stdin\n");
                exit_code = 1;
            }
            break;
        }
    }
    return exit_code;
}

File: zig/cat.zig
const std = @import("std");
pub fn main() u8 {
    var exit_code: u8 = 0;
    const stdout = std.io.getStdOut();
    const stdin = std.io.getStdIn();
    const stderr = std.io.getStdErr();

    var buffer : [1024]u8 = undefined;
    while (true) {
        const count = stdin.read(&buffer) catch |err| {
            stderr.writer().print("Failed to read from stdin: {}\n", .{err}) catch {};
            exit_code = 1;
            break;
        };
        if(count == 0) break;
        _ = stdout.write(buffer[0..count]) catch |err| {
            stderr.writer().print("Failed to write to stdout: {}\n", .{err}) catch {};
            exit_code = 1;
            break;
        };
    }
    return exit_code;
}
